### PR TITLE
[dv/doc] Add sec_cm_dv_framework

### PR DESCRIPTION
**Pull request by weicaiyang**
Signed-off-by: Weicai Yang <weicai@google.com>

### CODE REVIEW

File: doc/ug/sec_cm_dv_framework/index.md [hunk 1]
```diff
@@ -0,0 +1,128 @@
+---
+title: "Security Countermeasure Verification Framework"
+---
+
+## Purpose
+In a security chip, there are many various hardening security countermeasures implemented in security blocks.
```
Comment (matutem) [inline doc/ug/sec_cm_dv_framework/index.md:6]:
```suggestion
In a security chip, there are many hardening security countermeasures implemented in security blocks.
```

File: doc/ug/sec_cm_dv_framework/index.md [hunk 2]
```diff
@@ -0,0 +1,128 @@
...
+So, this may need to be tested one by one via forcing internal design signals on the security countermeasure related logic, which mimics attackers to flip the circuit.
+This kind of verification is needed but the effort could be tremendous.
+The work can be significantly reduced by using a unified verification framework.
+
+## Standardized Design Countermeasure Primitive
+Each countermeasure is implemented as a primitive - an RTL design building block written in SystemVerilog.
+The standardization of how these building blocks are designed paves the way for automated verification.
+
+There are 2 requirements that are needed in this primitive.
+1. Create a standard primitive for each kind of countermeasure, such as a security counter module, which can detect if the counter value is modified unintentionally.
```
Comment (matutem) [inline doc/ug/sec_cm_dv_framework/index.md:18]: This first sentence seems to repeat what you said above. Rework it?

File: doc/ug/sec_cm_dv_framework/index.md [hunk 3]
```diff
@@ -0,0 +1,128 @@
...
+These primitives could be reused across all blocks in the chip.
+Each primitive has an error indicator as an IO (or an internal signal if other blocks don’t need to connect with this error indicator).
+Once the design detects an attack, the error will be set.
+
+![diagram_of_sec_cm_prim](diagram_of_sec_cm_prim.svg)
+
+2. Once the error is set, the block should report a fatal alert or trigger unmasked interrupt to the system.
+There could be some other logics which are triggered when error occurs but not unified across the blocks.
```
Comment (matutem) [inline doc/ug/sec_cm_dv_framework/index.md:26]: "some other logics which are" -> "additional logic"?  Also, could you describe this in more detail? Or remove this mention? Perhaps you mean something along the line of custom additional logic, like cancelling some operations when errors occur.

File: doc/ug/sec_cm_dv_framework/index.md [hunk 4]
```diff
@@ -0,0 +1,128 @@
...
+![diagram_of_block_that_contains_sec_cm_primitive](diagram_of_block_that_contains_sec_cm_primitive.svg)
+
+## Embedded Common Checks In The Design
+At the top of each IP, an assertion check should be added to ensure that the fatal alert will be triggered once error is set.
+In the countermeasure primitive, there is another check to ensure that each instance of the countermeasure primitive always has this assertion check.
```
Comment (matutem) [inline doc/ug/sec_cm_dv_framework/index.md:31]: "always has this assertion check."... seems unclear, you mean something like "is connected to the fatal alert"?

File: doc/ug/sec_cm_dv_framework/index.md [hunk 5]
```diff
@@ -0,0 +1,128 @@
...
+
+  `ASSERT_PRIM_COUNT_ERROR_TRIGGER_ALERT(FsmCntAlertCheck_A, u_fsm.u_cnt, alert_tx_o[0])
+
+endmodule
+
+```
+
+## Verification Framework For The Standardized Design Countermeasures
+
+This verification framework includes 3 types of verification works, which are all needed in order to acheive the completed verification for countermeasures.
```
Comment (matutem) [inline doc/ug/sec_cm_dv_framework/index.md:74]:
```suggestion
This verification framework involves three different steps, which are all needed in order to acheive the completed verification for countermeasures.
```
Or something like that? I find it is better to use a word for numbers in descriptions, and the sentence is repeating "verification *work*", which makes it weaker.

File: doc/ug/sec_cm_dv_framework/index.md [hunk 6]
```diff
@@ -0,0 +1,128 @@
...
+### Primitive-Level Verification
+Use FPV or a simple simulation based testbench to verify the correctness of the error indicator as well as other functionality in the primitive.
+This unit-level testing eases the verification work at the higher level.
+Coverage closure for this primitive at higher-level can be eliminated.
+
+### FPV Proves The Embedded Assertion
+Use FPV to prove that the embedded assertion is true (fatal alert is triggered once error is set).
+This mathematically proves that an error unconditionally leads to the fatal alert.
+
+### Block-Level Verification With A Unified Semi-Automation Framework.
```
Comment (matutem) [inline doc/ug/sec_cm_dv_framework/index.md:85]:
```suggestion
### Block-Level Verification With A Unified Semi-Automated Framework
```

File: doc/ug/sec_cm_dv_framework/index.md [hunk 7]
```diff
@@ -0,0 +1,128 @@
...
+Triggering the fatal alert is usually not the only outcome of detecting an attack. Some other behavior may occur as well, such as wiping all the internal secrets, invaliding new commands.
```
Comment (matutem) [inline doc/ug/sec_cm_dv_framework/index.md:86]: This and some of the next sentences seem repeated. I think you are describing some specific part of the implementation, so state that more clearly...

File: doc/ug/sec_cm_dv_framework/index.md [hunk 8]
```diff
@@ -0,0 +1,128 @@
...
+```systemverilog
+foreach (sec_cm_pkg::sec_cm_if_proxy_q[i]) begin
+    sec_cm_pkg::sec_cm_base_if_proxy if_proxy = sec_cm_pkg::sec_cm_if_proxy_q[i];
+    inject_fault(if_proxy));  // call if_proxy.inject_fault
+    check_resp(if_proxy);
+end
+```
+This automation essentially relies on standardizing design countermeasure primitives.
+It uses the primitive types to anchor all the instances of countermeasure primitives and store all the handles that associate to the interface of the countermeasure primitive in a global location.
+The framework loops all the handles and provides a common sequence to check common behavior as well as callback functions that allow users to test the non-standard logics.
```
Comment (matutem) [inline doc/ug/sec_cm_dv_framework/index.md:126]:
```suggestion
The framework loops over all the handles and provides a common sequence to check common behavior as well as callback functions that allow users to test any non-standard behavior.
```

**Review by matutem — APPROVED** [review]
A few suggestions, and LGTM otherwise.

**Review by cindychip — APPROVED** [review]

File: doc/ug/sec_cm_dv_framework/index.md (hunk 4 above)
Comment (weicaiyang) [reply doc/ug/sec_cm_dv_framework/index.md]: I actually mean the instance of prim contains the error-trigger-alert assertion check.
Reworded a bit.

**Review by weicaiyang — COMMENTED** [review]
Thanks @matutem's comments. updated as suggested

**Pull request merged**